baseapp/baseapp.go

```diff
@@ -774,27 +792,51 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 // and DeliverTx. An error is returned if any single message fails or if a
 // Handler does not exist for a given message route. Otherwise, a reference to a
 // Result is returned. The caller must not commit state if an error is returned.
-func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*sdk.Result, error) {
+func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode, execGuarantee MsgExecutionGuaranteeMode) (*sdk.Result, error) {
 	msgLogs := make(sdk.ABCIMessageLogs, 0, len(msgs))
 	events := sdk.EmptyEvents()
 	var msgResponses []*codectypes.Any
 
+	txSuccess := false
+	msgCtx := ctx // This context may be replaced on each message for non-atomic execution.
+
 	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
 		if mode != runTxModeDeliver && mode != runTxModeSimulate {
 			break
 		}
 
+		// When dealing with non-atomic execution, we want each message to has its own context
+		var writeFn func()
+		if execGuarantee == NonAtomic {
+			msgCtx, writeFn = ctx.CacheContext()
+		}
+
 		handler := app.msgServiceRouter.Handler(msg)
 		if handler == nil {
 			return nil, errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "can't route message %+v", msg)
 		}
 
 		// ADR 031 request type routing
-		msgResult, err := handler(ctx, msg)
+		msgResult, err := handler(msgCtx, msg)
 		if err != nil {
+			if execGuarantee == NonAtomic {
+				value, err := codectypes.NewAnyWithValue(&tx.MsgFailureResponse{})
+				if err != nil {
+					return nil, err
+				}
+				msgResponses = append(msgResponses, value)
+				continue
+			}
 			return nil, errorsmod.Wrapf(err, "failed to execute message; message index: %d", i)
 		}
+		// at least one message was executed successfully
+		txSuccess = true
+
+		// Write the msg execution result to the context
+		if execGuarantee == NonAtomic {
+			writeFn()
+		}
```

Comment:
I figured non-atomic txs would be uncommon, so thought I'd save some context copying there (since the runTx is already caching the whole thing). Not a strong opinion though, happy to change if most people prefers that.

Reply:
Ah, I re-read your code, and you're creating a cached context also only on NonAtomic. Your small optimization makes sense then. Resolving.